### PR TITLE
Add error when document component isn't rendered

### LINKED ISSUE #10219: global css in _app throwing error
# Bug report

## Describe the bug

When importing a file from `_app.tsx`, an error is thrown and the page fails to render

>index.js:1 Error was not caught TypeError: Cannot read property 'parentNode' of null
>    at Object.push../src/pages/font.css.options.insert (font.css?2cd5:11)
>    at insertStyleElement (injectStylesIntoStyleTag.js:122)
>    at addStyle (injectStylesIntoStyleTag.js:219)
>    at addStylesToDom (injectStylesIntoStyleTag.js:94)
>    at push../node_modules/style->loader/dist/runtime/injectStylesIntoStyleTag.js.module.exports (injectStylesIntoStyleTag.js:251)
>    at Object../src/pages/font.css (font.css?2cd5:26)
>    at __webpack_require__ (bootstrap:792)
>    at fn (bootstrap:150)
>    at Module../src/pages/_app.tsx (with-apollo.tsx:36)
>    at __webpack_require__ (bootstrap:792)

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

a relatively simple `_app.js`:

```tsx
import * as React from 'react';
import { Global } from '@emotion/core';
import { Monterey } from '@tedconf/monterey';
import App from 'next/app';
import { ApolloProvider } from '@apollo/react-hooks';
import { I18nProvider } from '@lingui/react';
import objectFitImages from 'object-fit-images';
import withApollo from '../lib/with-apollo';
import './font.css';

const catalogEN = require('../locales/en/messages.po');

class MyApp extends App<{ apollo: any }> {
  componentDidMount() {
    objectFitImages();
  }

  render() {
    const { Component, pageProps, apollo } = this.props;

    return (
      <I18nProvider language="en" catalogs={{ en: catalogEN }}>
        <ApolloProvider client={apollo}>
          <link rel="icon" href="/favicon.ico" />
          <Global
            styles={{
              html: {
                fontFamily: 'Inter, sans-serif',
              },
            }}
          />
          <Monterey>
            <Component {...pageProps} />
          </Monterey>
        </ApolloProvider>
      </I18nProvider>
    );
  }
}

export default withApollo(MyApp);
```

and a simple css file:

```css
@font-face {
  font-family: 'Inter';
  font-weight: 100 900;
  font-style: normal;
  font-named-instance: 'Regular';
  font-display: swap;
  src: url('/fonts/inter-upright.var.woff2') format('woff2');
}
@font-face {
  font-family: 'Inter';
  font-weight: 100 900;
  font-style: italic;
  font-named-instance: 'Italic';
  font-display: swap;
  src: url('/fonts/inter-italic.var.woff2') format('woff2');
}
body {
  font-variant-ligatures: normal;
}
```

## Expected behavior

The CSS should load or the error should be caught and decipherable

## System information

- OS: macOS
- Browser (if applies) chrome
- Version of Next.js: 9.0.2

## Additional context

Just tried to switch from `withCSS`.

### PR DESCRIPTION
**Pull request by ijjk**
If a custom `_document` is added but not all of the expected document components are rendered it can cause unintended errors as noticed in https://github.com/vercel/next.js/issues/10219 so this adds detecting when one of the expected document components isn't rendered and shows an error. 

Closes: https://github.com/vercel/next.js/issues/10219

### CODE REVIEW

**Review by Timer — APPROVED** [review]
Epic!

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js error/missing-document-component | Change |
| - | - | - | - |
| buildDuration | 12s | 12.2s | ⚠️ +235ms |
| nodeModulesSize | 57.5 MB | 57.5 MB | ⚠️ +3.48 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | ijjk/next.js error/missing-document-component | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.103 | 2.182 | ⚠️ +0.08 |
| / avg req/sec | 1188.96 | 1145.85 | ⚠️ -43.11 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.223 | 1.178 | -0.05 |
| /error-in-render avg req/sec | 2043.76 | 2122.88 | +79.12 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js error/missing-document-component | Change |
| - | - | - | - |
| 677f882d2ed8..7f47.js gzip | 10.3 kB | 10.3 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-978aff6..e7ab.js gzip | 7.2 kB | 7.2 kB | ✓ |
| webpack-e067..f178.js gzip | 751 B | 751 B | ✓ |
| Overall change | 57.2 kB | 57.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js error/missing-document-component | Change |
| - | - | - | - |
| 677f882d2ed8..dule.js gzip | 6.15 kB | 6.15 kB | ✓ |
| framework.HA..dule.js gzip | 39 kB | 39 kB | ✓ |
| main-46b15db..dule.js gzip | 6.24 kB | 6.24 kB | ✓ |
| webpack-07c5..dule.js gzip | 751 B | 751 B | ✓ |
| Overall change | 52.1 kB | 52.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js error/missing-document-component | Change |
| - | - | - | - |
| polyfills-4b..e242.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js error/missing-document-component | Change |
| - | - | - | - |
| _app-9a0b9e1..b37e.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-28298..e0c9.js gzip | 3.44 kB | 3.44 kB | ✓ |
| hooks-89731c..c609.js gzip | 887 B | 887 B | ✓ |
| index-17468f..5d83.js gzip | 227 B | 227 B | ✓ |
| link-000f151..65d4.js gzip | 1.29 kB | 1.29 kB | ✓ |
| routerDirect..924c.js gzip | 284 B | 284 B | ✓ |
| withRouter-7..c13d.js gzip | 284 B | 284 B | ✓ |
| Overall change | 7.69 kB | 7.69 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js error/missing-document-component | Change |
| - | - | - | - |
| _app-75d3a82..dule.js gzip | 625 B | 625 B | ✓ |
| _error-65c8a..dule.js gzip | 2.29 kB | 2.29 kB | ✓ |
| hooks-cbf13f..dule.js gzip | 387 B | 387 B | ✓ |
| index-b9a643..dule.js gzip | 226 B | 226 B | ✓ |
| link-4cfda7a..dule.js gzip | 1.26 kB | 1.26 kB | ✓ |
| routerDirect..dule.js gzip | 284 B | 284 B | ✓ |
| withRouter-f..dule.js gzip | 282 B | 282 B | ✓ |
| Overall change | 5.35 kB | 5.35 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js error/missing-document-component | Change |
| - | - | - | - |
| _buildManifest.js gzip | 322 B | 322 B | ✓ |
| _buildManife..dule.js gzip | 330 B | 330 B | ✓ |
| Overall change | 652 B | 652 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js error/missing-document-component | Change |
| - | - | - | - |
| index.html gzip | 949 B | 949 B | ✓ |
| link.html gzip | 955 B | 955 B | ✓ |
| withRouter.html gzip | 940 B | 940 B | ✓ |
| Overall change | 2.84 kB | 2.84 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js error/missing-document-component | Change |
| - | - | - | - |
| buildDuration | 13.3s | 13.2s | -118ms |
| nodeModulesSize | 57.5 MB | 57.5 MB | ⚠️ +3.48 kB |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js error/missing-document-component | Change |
| - | - | - | - |
| 677f882d2ed8..7f47.js gzip | 10.3 kB | 10.3 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-978aff6..e7ab.js gzip | 7.2 kB | 7.2 kB | ✓ |
| webpack-e067..f178.js gzip | 751 B | 751 B | ✓ |
| Overall change | 57.2 kB | 57.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js error/missing-document-component | Change |
| - | - | - | - |
| 677f882d2ed8..dule.js gzip | 6.15 kB | 6.15 kB | ✓ |
| framework.HA..dule.js gzip | 39 kB | 39 kB | ✓ |
| main-46b15db..dule.js gzip | 6.24 kB | 6.24 kB | ✓ |
| webpack-07c5..dule.js gzip | 751 B | 751 B | ✓ |
| Overall change | 52.1 kB | 52.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js error/missing-document-component | Change |
| - | - | - | - |
| polyfills-4b..e242.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js error/missing-document-component | Change |
| - | - | - | - |
| _app-9a0b9e1..b37e.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-28298..e0c9.js gzip | 3.44 kB | 3.44 kB | ✓ |
| hooks-89731c..c609.js gzip | 887 B | 887 B | ✓ |
| index-17468f..5d83.js gzip | 227 B | 227 B | ✓ |
| link-000f151..65d4.js gzip | 1.29 kB | 1.29 kB | ✓ |
| routerDirect..924c.js gzip | 284 B | 284 B | ✓ |
| withRouter-7..c13d.js gzip | 284 B | 284 B | ✓ |
| Overall change | 7.69 kB | 7.69 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js error/missing-document-component | Change |
| - | - | - | - |
| _app-75d3a82..dule.js gzip | 625 B | 625 B | ✓ |
| _error-65c8a..dule.js gzip | 2.29 kB | 2.29 kB | ✓ |
| hooks-cbf13f..dule.js gzip | 387 B | 387 B | ✓ |
| index-b9a643..dule.js gzip | 226 B | 226 B | ✓ |
| link-4cfda7a..dule.js gzip | 1.26 kB | 1.26 kB | ✓ |
| routerDirect..dule.js gzip | 284 B | 284 B | ✓ |
| withRouter-f..dule.js gzip | 282 B | 282 B | ✓ |
| Overall change | 5.35 kB | 5.35 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js error/missing-document-component | Change |
| - | - | - | - |
| _buildManifest.js gzip | 322 B | 322 B | ✓ |
| _buildManife..dule.js gzip | 330 B | 330 B | ✓ |
| Overall change | 652 B | 652 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js error/missing-document-component | Change |
| - | - | - | - |
| _error.js | 1.03 MB | 1.03 MB | ⚠️ +362 B |
| 404.html | 4.18 kB | 4.18 kB | ✓ |
| hooks.html | 3.82 kB | 3.82 kB | ✓ |
| index.js | 1.03 MB | 1.03 MB | ⚠️ +362 B |
| link.js | 1.07 MB | 1.07 MB | ⚠️ +362 B |
| routerDirect.js | 1.07 MB | 1.07 MB | ⚠️ +362 B |
| withRouter.js | 1.07 MB | 1.07 MB | ⚠️ +362 B |
| Overall change | 5.28 MB | 5.28 MB | ⚠️ +1.81 kB |

</details>

</details>
Commit: 8aaf216e6ffa0334d37fd8ba4c883fbaca475331

File: packages/next/next-server/server/render.tsx [hunk 1]
```diff
@@ -787,6 +794,29 @@ export async function renderToHTML(
...
+    const plural = nonRenderedComponents.length !== 1 ? 's' : ''
+
+    if (nonRenderedComponents.length) {
+      console.warn(
+        `Expected Document Component${plural} ${nonRenderedComponents.join(
+          ', '
+        )} ${
+          plural ? 'were' : 'was'
+        } not rendered. Make sure you render them in your custom \`_document\`\n` +
+          `See more info here https://err.sh/next.js/missing-document-component`
```
Comment (pruhstal) [inline packages/next/next-server/server/render.tsx:815]: @kodiakhq just got this error and wanted to point out that this link 404s.

Comment (ijjk) [reply packages/next/next-server/server/render.tsx:815]: Hi, this link should be available now, thanks for pointing that out

**Pull request merged**